zerver/migrations/0542_move_emoji_images.py

```diff
@@ -0,0 +1,301 @@
+                CopySource=f"{settings.S3_AVATAR_BUCKET}/{base_path}/{old_file_name}.original",
+                MetadataDirective="REPLACE",
+                Metadata=metadata,
+                ContentDisposition=old_data["ContentDisposition"],
+                ContentType=old_data["ContentType"],
+                CacheControl="public, max-age=31536000, immutable",
+            )
+
+            content_type = guess_type(new_file_name)[0]
+            if content_type is None or not content_type.startswith("image/"):
+                raise SkipImageError(f"Content is unknown or not an image: {content_type}")
+            animated, still = resize_emoji(original_bytes, new_file_name)
+            try:
+                avatar_bucket.Object(f"{base_path}/{new_file_name}").put(
+                    Metadata=metadata,
+                    ContentType=content_type,
+                    CacheControl="public, max-age=31536000, immutable",
+                    Body=animated,
+                )
+                if still is not None:
+                    filename_no_extension = os.path.splitext(new_file_name)[0]
+                    avatar_bucket.Object(f"{base_path}/still/{filename_no_extension}.png").put(
+                        Metadata=metadata,
+                        ContentType="image/png",
+                        CacheControl="public, max-age=31536000, immutable",
+                        Body=still,
+                    )
+            except botocore.exceptions.ClientError as e:
+                raise SkipImageError(f"Failed to upload new file: {e}")
+
+            emoji.file_name = new_file_name
+            emoji.save(update_fields=["file_name"])
+
+            for path in to_delete:
+                with contextlib.suppress(botocore.exceptions.ClientError):
+                    avatar_bucket.Object(f"{base_path}/{path}").delete()
+        except SkipImageError as e:
+            logging.error(
+                "Failed to re-thumbnail %s/emoji/images/%s: %s", emoji.realm_id, emoji.file_name, e
+            )
+
+
+def used_emoji_filenames(apps: StateApps, realm: Any) -> Set[str]:
+    # We can remove the old filename for any emoji that is _not_
+    # referenced in the rendered body of a message; re-rending is
+    # complicated and time-consuming, and email digests exist
+    # referencing the URLs regardless.
+    Message = apps.get_model("zerver", "Message")
+    BATCH_SIZE = 1000
+    seen = set()
+    min_id = 0
+    while True:
+        messages = Message.objects.filter(
+            realm=realm,
+            id__gt=min_id,
+            rendered_content__contains=f"/user_avatars/{realm.id}/emoji/images/",
+        )[:BATCH_SIZE]
+
+        for message in messages:
+            for match in re.findall(
+                rf'/user_avatars/{realm.id}/emoji/images/([^"]+)', message.rendered_content
+            ):
+                seen.add(match)
+
+        if len(messages) < BATCH_SIZE:
+            break
+        min_id = messages[-1]
```

Comment:
You probably want `.id` here, to avoid the type changing from integer to `MEssage` after the first iteration.